### ISSUE TITLE
Operation not supported error on setting acls with Linux Kernel 6.2-rc4

### ISSUE THREAD

**Issue by satmandu** [openzfs/zfs#14403]
Fixed by #14415 

### System information

Type | Version/Name
 --- | ---
Distribution Name	| Ubuntu
Distribution Version	| 22.10 
Kernel Version	| 6.2-rc4
Architecture	| x86_64
OpenZFS Version	| [2.1.99/master](https://github.com/openzfs/zfs/commit/a379083d9f2bb9dd80f4636e593bcb2c1d94d11b) 

### Describe the problem you're observing

This fails with kernel 6.2rc4:
```
touch a
sudo setfacl -m u:root:r a
setfacl: a: Operation not supported
```
It works with kernel 6.1.6:
```
satadru@cheekon ~/tmp$ touch a
sudo setfacl -m u:root:r a
satadru@cheekon ~/tmp$
```

### Include any warning/errors/backtraces from the system logs
There is noting else in the log. This problem also occurs when doing a `cp -a` for any files with acls already set.

(This was self-built from master, so it's also possible I screwed something up with the build... but it does work fine on 6.1.6...)

**Comment by satmandu** [comment]
(I am not sure if the acl changes in 6.2 are implicated: https://lore.kernel.org/lkml/20221212111919.98855-1-brauner@kernel.org/ )

**Comment by satmandu** [comment]
Relevant commit is https://github.com/torvalds/linux/commit/6a518afcc2066732e6c5c24281ce017bbbd85506

**Comment by satmandu** [comment]
@ckane thoughts?

**Comment by ckane** [comment]
Thanks for the report. Likely regression introduced either by my changes or not completely chasing the upstream kernel changes that broke it in the first place. I'll take a look but may not have time to fix until this weekend. Your example could probably stand to be incorporated into the test suite, as well.

**Comment by ckane** [comment]
Also gonna tag @behlendorf

**Comment by ckane** [comment]
So, it is *possible* that the changes might have left something out, but the major change in the FACL code being chased in the linked PR is that `get_acl()` was supplanted by `get_inode_acl()`, but I see no corresponding change to the `set_acl()` code in my Linux 6.2-rc4, so I don't see any obvious changes that need to similarly be applied to the "set" counterpart (could use someone's help identifying this).
Are you using the same version of the ZFS code in both your 6.1.6 and 6.2rc4 kernels?
Can you should the `ls -l` output for both, after the ACL change, to verify that the old kernel isn't just silently failing?

There have also been some recent changes to `module/os/linux/zfs/zfs_acl.c` and `module/os/linux/zfs/zfs_vnops.c`, but I don't see anything in either of those sets that should cause this behavior.

Running `configure` against latest `master` indicates that both of the tests used to identify which version of `set_acl` to use are failing, so this might be the culprit:
```c
/* #undef HAVE_SET_ACL */
/* #undef HAVE_SET_ACL_USERNS */
```

**Comment by ckane** [comment]
```
zfs-test/build/inode_operations_set_acl_userns/inode_operations_set_acl_userns.c:96:36:
error: initialization of 
‘int (*)(struct user_namespace *, struct dentry *, struct posix_acl *, int)’ from incompatible pointer type
‘int (*)(struct user_namespace *, struct inode *, struct posix_acl *, int)’
[-Werror=incompatible-pointer-types]
```

Looks like the second argument to `set_acl` was changed to `struct dentry *` at some point, while we are still expecting it to be a `struct inode*`

**Comment by ckane** [comment]
@satmandu I don't actually use ACLs on my system, so I am not certain if it was working before or not. However, I made some changes that re-connect the `zpl_set_acl` implementation to the FS operations in  #14415. It doesn't seem to change the behavior on my system, but I'd like to know if it does on yours. I still think I'm missing something, but the commit I made in that PR should be a step in the right direction.

Would love some advice from @behlendorf or others about what else might be being overlooked.
